### PR TITLE
Add target selection and tab completion by name

### PR DESCRIPTION
**Pull request by wvu**
# ~WIP~

This patch looks way simpler than it should be. Be careful.

- [x] `show targets` and see all targets
- [x] `options` and see selected target
- [x] `set target` and show tab completions
- [x]  Tab-complete a target by name
- [x] `options` and see that it's selected
- [x] Try the same with an index

~Yes, I'm aware that tab completion is breaking on spaces and not natural-sorted.~

- The former appears to be a design limitation of our Readline implementation. We basically need a way to prevent word splitting - so quoting or escaping.
- I'm not sure what to do about the latter yet. I tracked this one down to Readline itself, and we hacked on it a bit but couldn't get it to work beyond significant monkeypatching.

### CODE REVIEW

**Comment by busterb** [comment]
There are still a few things that reference datastore['TARGET'] that might need some cleanup too.

File: lib/msf/core/exploit.rb [hunk 1]
```diff
@@ -674,13 +674,16 @@ def target
   # The target index that has been selected.
   #
   def target_index
-    target_idx = datastore['TARGET']
+    target_idx = Integer(datastore['TARGET']) rescue datastore['TARGET']
```
[the remark this answers is not in the document]
Comment (wvu-r7) [reply lib/msf/core/exploit.rb]: FYI, I rewrote the patch to use `target_idx` exclusively. Previously I had set `target_idx_or_name`. I think I might revert back to that for additional clarity.

File: lib/msf/core/exploit.rb [hunk 2]
```diff
@@ -674,13 +674,16 @@ def target
...
 
     default_idx = default_target || 0
     # Use the default target if one was not supplied.
     if (target_idx == nil and default_idx and default_idx >= 0)
       target_idx = default_idx
+    elsif target_idx.is_a?(String)
```
[the remark this answers is not in the document]
Comment (wvu-r7) [reply lib/msf/core/exploit.rb]: I'm pretty sure `NilClass` and `String` are mutually exclusive here. I didn't have any problems during testing. Did I miss an edge case when I refactored the patch? D:

**Comment by busterb** [comment]
Recording for posterity, this sadly did not work:

```
diff --git a/lib/rex/ui/text/dispatcher_shell.rb b/lib/rex/ui/text/dispatcher_shell.rb
index 24096284bf..3c35cb5cd4 100644
--- a/lib/rex/ui/text/dispatcher_shell.rb
+++ b/lib/rex/ui/text/dispatcher_shell.rb
@@ -320,6 +320,15 @@ module DispatcherShell
     self.on_command_proc = nil
   end
 
+  class NaturalArray < Array
+    def sort
+      sort_by {|e| e.split(/(\d+)/).map {|a| a =~ /\d+/ ? a.to_i : a }}
+    end
+    def sort!
+      sort_by! {|e| e.split(/(\d+)/).map {|a| a =~ /\d+/ ? a.to_i : a }}
+    end
+  end
+
   #
   # This method accepts the entire line of text from the Readline
   # routine, stores all completed words, and passes the partial
@@ -342,7 +351,7 @@ module DispatcherShell
     self.tab_words = str_words
 
     # Pop the last word and pass it to the real method
-    tab_complete_stub(self.tab_words.pop)
+    NaturalArray.new(tab_complete_stub(self.tab_words.pop))
   end
```

**Comment by wvu-r7** [comment]
# Release Notes

This implements the ability to select an exploit module's target by name instead of just index. Tab completion for target names has also been added.

**Pull request merged**